HDFEventOutputer.cc

```diff
@@ -0,0 +1,204 @@
+  auto& laneSerializers = serializers_[iLaneIndex];
+  auto group = iCallback.group();
+  laneSerializers[iDataProduct.index()].doWorkAsync(*group, iDataProduct.address(), std::move(iCallback));
+}
+
+void HDFEventOutputer::outputAsync(unsigned int iLaneIndex, EventIdentifier const& iEventID, TaskHolder iCallback) const {
+  auto start = std::chrono::high_resolution_clock::now();
+  auto [offsets, buffer] = writeDataProductsToOutputBuffer(serializers_[iLaneIndex]);
+  queue_.push(*iCallback.group(), [this, iEventID, iLaneIndex, callback=std::move(iCallback), buffer = std::move(buffer), offsets = std::move(offsets)]() mutable {
+      auto start = std::chrono::high_resolution_clock::now();
+      const_cast<HDFEventOutputer*>(this)->output(iEventID, serializers_[iLaneIndex], std::move(buffer), std::move(offsets));
+        serialTime_ += std::chrono::duration_cast<decltype(serialTime_)>(std::chrono::high_resolution_clock::now() - start);
+      callback.doneWaiting();
+    });
+    auto time = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::high_resolution_clock::now() - start);
+    parallelTime_ += time.count();
+}
+
+void HDFEventOutputer::printSummary() const  {
+  std::cout <<"HDFEventOutputer\n  total serial time at end event: "<<serialTime_.count()<<"us\n"
+    "  total parallel time at end event: "<<parallelTime_.load()<<"us\n";
+  summarize_serializers(serializers_);
+}
+
+
+
+void 
+HDFEventOutputer::output(EventIdentifier const& iEventID, 
+                         SerializeStrategy const& iSerializers,
+                         std::vector<char> iBuffer,
+                         std::vector<uint32_t> iOffsets ) {
+  hdf5::Group gid = hdf5::Group::open(file_, "Lumi");   
+  if (firstEvent_) {
+    firstEvent_ = false;
+     auto r = hdf5::Attribute::open(gid, "run");
+     r.write(iEventID.run);  
+     auto sr = hdf5::Attribute::open(gid, "lumisec");
+     sr.write(iEventID.lumi);  
+  }
+  eventID_ = iEventID;
+  std::vector<unsigned long long> ids;
+  ids.push_back(iEventID.event);
+  write_ds<unsigned long long>(gid, "Event_IDs", ids);
+  write_ds<char>(gid, "products", iBuffer);
+  std::vector<uint32_t> offset;
+  offset.resize(iOffsets.size()); 
+ // std::copy( iOffsets.begin(), iOffsets.end(), offset.begin() );
+  write_ds<uint32_t>(gid, "offsets", iOffsets); 
+}
+
+void 
+HDFEventOutputer::writeFileHeader(SerializeStrategy const& iSerializers) {
+  constexpr hsize_t ndims = 1;
+  constexpr hsize_t     dims[ndims] = {0};
+  constexpr hsize_t     chunk_dims[ndims] = {128};
+  constexpr hsize_t     max_dims[ndims] = {H5S_UNLIMITED};
+  auto space = hdf5::Dataspace::create_simple (ndims, dims, max_dims); 
+  auto prop   = hdf5::Property::create();
+  prop.set_chunk(ndims, chunk_dims);
+  hdf5::Group g = hdf5::Group::create(file_, "Lumi");
+  hdf5::Dataset::create<int>(g, "Event_IDs", space, prop);
```

Comment:
Given the dataset names appear multiple places in the code, how about making them `constexpr const char* const` named variables in the anonymous namespace?